File: rust/src/solutions/s0009.rs
use proconio::input;
use std::time::Instant;

// Solution for Project Euler problem 9
// Copyright michimani All rights reserved.
//
// https://projecteuler.net/problem=9
pub fn solve() {
    let org = 1000;
    println!("(original: {})", org);
    input! {
      n: usize,
    }

    let start = Instant::now();

    let mut answer = 0;

    for a in 1..(n / 3) {
        for b in (a + 1)..(n / 2) {
            let c = n - a - b;
            if a.pow(2) + b.pow(2) == c.pow(2) {
                answer = a * b * c;
                break;
            }
        }
    }

    println!("answer is {}", answer);

    let end = start.elapsed();
    println!(
        "\nIt took {}.{:03} seconds.",
        end.as_secs(),
        end.subsec_nanos() / 1_000_000
    );
}

File: python/solution/s0009.py
#
# Solution for Project Euler problem 9
# Copyright michimani All rights reserved.
#
# https://projecteuler.net/problem=9
#

import math


def solve():
    is_ok = False
    a = 1
    b = 2
    c = 3
    max_c = 1000 - a - b
    sq_list = [num**2 for num in range(3, max_c + 1)]
    for c_sq in sq_list:
        for a_sq in range(1, math.ceil(c_sq / 2)):
            if is_nature_sqrt(a_sq) is True:
                b_sq = c_sq - a_sq
                if is_nature_sqrt(b_sq) is True:
                    a = round(a_sq ** (1 / 2))
                    b = round(b_sq ** (1 / 2))
                    c = round(c_sq ** (1 / 2))
                    if a + b + c == 1000:
                        is_ok = True
                        break

        if is_ok is True:
            break

    return a * b * c


def is_nature_sqrt(num):
    if num > 0:
        return num == math.floor(num ** (1 / 2)) ** 2
    else:
        return False


if __name__ == "__main__":
    print(solve())
